editor/src/components/canvas/controls/grid-controls.tsx

```diff
@@ -2163,10 +2175,269 @@ function useIsGridItemInteractionActive() {
   )
 }
 
-function useIsGridItemSelected() {
+function useSelectedGridItems(): ElementPath[] {
   const selectedViewsRef = useRefEditorState((store) => store.editor.selectedViews)
   const jsxMetadataRef = useRefEditorState((store) => store.editor.jsxMetadata)
-  return selectedViewsRef.current.some((selected) =>
+  return selectedViewsRef.current.filter((selected) =>
     MetadataUtils.isGridItem(jsxMetadataRef.current, selected),
   )
 }
+
+const rulerMarkerIconSize = 12 // px
+
+type RulerMarkerData = {
+  columnStart: RulerMarkerPositionData
+  columnEnd: RulerMarkerPositionData
+  rowStart: RulerMarkerPositionData
+  rowEnd: RulerMarkerPositionData
+}
+
+type RulerMarkerPositionData = {
+  top: number
+  left: number
+  position: GridPositionOrSpan | null
+  counterpart: GridPositionOrSpan | null
+  bound: 'start' | 'end'
+}
+
+const RulerMarkers = React.memo((props: { path: ElementPath }) => {
+  const markers: RulerMarkerData | null = useEditorState(
+    Substores.metadata,
+    (store) => {
+      const elementMetadata = MetadataUtils.findElementByElementPath(
+        store.editor.jsxMetadata,
+        props.path,
+      )
+      if (elementMetadata == null) {
+        return null
+      }
+
+      const elementGridProperties = elementMetadata.specialSizeMeasurements.elementGridProperties
+      if (elementGridProperties == null) {
+        return null
+      }
+
+      const originalGrid = findOriginalGrid(store.editor.jsxMetadata, EP.parentPath(props.path))
+      if (originalGrid == null) {
+        return null
+      }
+
+      const parentGridCellGlobalFrames =
+        elementMetadata.specialSizeMeasurements.parentGridCellGlobalFrames
+      if (parentGridCellGlobalFrames == null) {
+        return null
+      }
+
+      const cellBounds = getGridChildCellCoordBoundsFromCanvas(
+        elementMetadata,
+        parentGridCellGlobalFrames,
+      )
+      if (cellBounds == null) {
+        return null
+      }
+
+      const cellBoundsColumnIndex = cellBounds.column - 1
+      const firstRow = parentGridCellGlobalFrames[0]
```

Comment:
This doesn't look safe for the empty array case.

Reply:
[`c901843` (#6654)](https://github.com/concrete-utopia/utopia/pull/6654/commits/c90184380120304d2d15e947cd7f3e2c41c0fcbe) (I think it's fairly unlikely to have an empty array, but added a guard 👍 )